Step2 - 문자열 덧셈 계산기

Mo-bile opened the pull request:
안녕하세요?
문자열 덧셈 계산기 리뷰 요청드립니다.

아까 라이브 강의에서 주신 가이드를 떠올리며 비슷하면서 리팩토링 했습니다.

1. 테스트 코드 작성도 직접하면 좋은데 시간이 늦어서 못한건 아쉬운데 금번 PR 피드백 이후에 직접 작성해서 추가해도 괜찮을까요?
2. 메소드 분리를 할 때 분리한 메소드의 매개변수 수의 제한같은 것이 있을까요?
아니면 무조건 몇개 제한 보다, 어떤 경우에는 어떤 변수는 매개변수로 넘겨도 되고, 어떤경우에는 넘기면 안되는 이런것이 있을까요?

감사합니다 ☺️

Commit:
add dependency junit-jupiter-params

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Mo-bile committed:
feat : 문자열 덧셈 계산기 구현 완료

Mo-bile committed:
refactor
1. split() 부분 함수 분리
2. 문자열에서 정수로 변경 부분 분리

Mo-bile committed:
refactor : null, empty 조건문 메소드 분리

Mo-bile committed:
test : 기본제공 테스트 추가

javajigi commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,38 @@
+
+public class StringAddCalculator {
+
+    public static int splitAndSum(String text) {
```
강의 내용을 그대로 따라하지 않은 것이 인상적임 👍

javajigi commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    private static String isExist(String text) {
+        return (text == null || text.isEmpty()) == true ? "0" : text;
```
오호.
신선한 접근임
단, 메서드 이름에 isExist이라면 반환값은 boolean을 기대하는 것이 일반적인 관례인데 String이라 혼란스러움이 있지 않을까?

javajigi commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    private static String[] splitText(String text) {
+        if (text.contains(",") || (text.contains(":"))) {
```
거꾸로 접근한 것도 👍

javajigi commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,38 @@
+    private static int parseToInt(String token) {
+        int anInt = Integer.parseInt(token);
+        if (anInt < 0) {
+            throw new RuntimeException();
```
에러 메시지를 추가하는 습관을 들여도 좋지 않을까?

javajigi commented on src/test/java/study/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,46 @@
+
+class StringAddCalculatorTest {
+    @Test
+    public void splitAndSum_null_또는_빈문자() {
```
어제 강의에서 AI가 추천해준 @NullAndEmptySource 한번 적용해 보면 어떨까?

javajigi reviewed (changes requested):
어제 강의를 봤으면 그대로 따라할 것 같은데요.
재영님만의 스타일로 다시 구현한 점이 인상적이네요. 👍 
바로 merge하기 보다는 다음 단계 연습을 해보면 어떨까요?

1. StringAddCalculator.java의 프로덕션 코드를 모두 삭제
2. StringAddCalculatorTest.java는 현재 코드를 그대로 유지
3. 실패하는 test case를 하나씩 성공 시키기 위해 아무 것도 없는 StringAddCalculator.java 을 하나씩 구현해 간다.
4. test case를 하나씩 성공 시키기 단계마다 리팩터링을 진행한다.

작게라도 TDD 사이클을 경험해 보면 어떨까 제안해 봅니다.

javajigi commented:
Q: 테스트 코드 작성도 직접하면 좋은데 시간이 늦어서 못한건 아쉬운데 금번 PR 피드백 이후에 직접 작성해서 추가해도 괜찮을까요?
- 제가 앞의 피드백에 남긴 연습을 한다면 굳이 테스트 코드 작성을 이곳에서 연습하지 않아도 될 것 같아요.
- 테스트 코드를 직접 작성하는 연습은 3단계에서 해보면 어떨까요?

Q:  메소드 분리를 할 때 분리한 메소드의 매개변수 수의 제한같은 것이 있을까요?
 아니면 무조건 몇개 제한 보다, 어떤 경우에는 어떤 변수는 매개변수로 넘겨도 되고, 어떤경우에는 넘기면 안되는 이런것이 있을까요?
- 클린 코드 책을 보면 다음과 같은 내용이 있어요.
- 저도 교육 중에 아래 내용으로 가이드하고 있어요. 가능하면 2개까지 허용하는 방식으로 도전해 볼 것을 추천합니다.

```
## 함수 인수
- 함수에서 이상적인 인수 개수는 0개(무항)이다. 다음은 1개이고, 다음은 2개이다.
- 3개는 가능한 피하는 편이 좋다.
- 4개 이상은 특별한 이유가 있어도 사용하면 안된다
```

Commit:
refactor
1. 상수 리팩토링
2. String 반환이므로 메소드명 isExist 를 defaultIfEmpty 로 메소드명 변경
3. 숫자만 단독으로 입력시  split() 메소드로 그대로 return 되어서 splitText 메소드 내 불필요한 NORMAL_REGEX 분기 제거

Commit:
test :
모든 테스트 케이스 red

Commit:
feat : null_또는_빈문자 테스트 통과 기능 개발 완료

Commit:
refactor : null_또는_빈문자 통과 기능 함수 분리

Commit:
feat : 숫자하나 테스트 통과 기능 개발 완료

Commit:
refactor : 숫자하나 테스트
메서드 분리, 코드라인 축소

Commit:
feat : 쉼표구분자 테스트 통과 기능 개발 완료

Commit:
refactor : 쉼표구분자 테스트 리팩토링
1. 문자열 분리 메서드 분리
2. 문자열 배열을 정수 배열로 분로 변환 분리
3. adder() 메서드 분리하여 연산 분리

Commit:
feat : 쉼표_또는_콜론_구분자 테스트 통과 완료

Commit:
refacotr : 쉼표_또는_콜론_구분자 테스트 리팩터링
상수 분리

Commit:
feat : custom_구분자 테스트 통과 완료

Commit:
refactor : custom_구분자 테스트
1. 상수 분리
2. splitText 메서드 내부 축소

Commit:
feat : negative 수 탐지 테스트 통과

Commit:
refactor : negative 수 탐지 테스트
1. 음수 탐지 역할을 하는 메서드 분리
(컴포즈 메서드 패턴 위반...)

Mo-bile commented:
안녕하세요? 요청주신 바 대로 '작게라도 TDD 사이클'을 경험했습니다. 그 결과 흥미로운 경험을 하게 되었네요 😊👍

금번 과정을 하면서 새로운 형태의 코드가 탄생한 것 같으면서도 강의 중에 보여주셨던 코드와 비슷해진 느낌이 들어요

몇가지 질문드립니다

1. 커밋의 기준은 어떻게 나누면 될까요?
- 금번에 TDD 싸이클을 경험하면서 다음과 같이 커밋했습니다.
```
예)
우선 기존 프로덕트 코드를 모두 제거(레드) -> 커밋
테스트 하나 통과(그린) -> 커밋
통과한 프로덕션 코드 리팩터링 -> 커밋

즉 "레드 -> 그린 -> 리팩터링 -> 그린 -> 리팩터링 -> 그린 -> 리팩터링 -> 그린 -> 리팩터링"  순환
```
커밋하면서 느낀건 아 커밋이 너무 많은데? 적정 TDD 싸이클에서 적정커밋은 무엇일까 궁금해졌습니다.

2. 삼항연산자를 리팩터링에 사용해도 될까요?
- 만약 그러하다면 삼항연산자 참과 거짓의 기준을 나누는 메서드(제 코드에서는 'isNullOrEmpty' 메서드)는 이 역할만 집중해서 변수명을 정하면 되는 것 맞을까요?
- 생각이 든 것은 삼항연산자의 결과에 대한 '참/거짓' 부분의 내용을 메서드 명에 내포해도 되지 않을까 이런 생각이 들었습니다.
- 추가로 든 생각은 무언가 삼항연산자 사용이 indent depth 를 줄이려는 꼼수 같다는 생각도 들었습니다. 이 부분은 어떻게 생각하실까요?

3. 리팩터링 한 기능에 새로운 기능을 구현하면서 깨지는것이 당연한 것 맞을까요?
우선 TDD 싸이클을 겪으면서 느낀 소회를 가볍게 말씀드리면
기능을 만들고 리팩터링하고, 그 다음 순서의 기능을 구현 만들 때 깔끔해진 리팩터링 된 코드가 깨질때 내심 마음이 아팠습니다. 그러면서도 그 구현한 기능을 다시 만들면서 다행이라고 생각들었습니다. 이런 경험이 TDD 싸이클에서 당연한 것일까요?

4. 음수 예외처리 기능을 리팩터링 하는데 어려웠습니다.
- 인덴트 뎁스를 1로 맞추고 싶어서 많은 고민을 했는데 입력값의 음수 확인을 위해서는 결국 '반복문' + '조건문'을 피할수가 없더라고요.
- 스트림을 사용도 시도했지만 인덴트 뎁스 1을 맞출수 없었고요...
- 그래서 결국에 컴포즈 매서드 패턴을 위반하는 방식으로 분리한 메서드 안에서 메서드를 분리했습니다.
- 이 상황에서는 이 방식이 최선일까요?

혹시 이해가 어려우신 내용이 있으면 언제든 부연질문 주시면 감사하겠습니다
긴 내용 읽어주시고 리뷰해주셔서 감사드립니다 👍

javajigi commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,48 @@
+    public static final String CUSTOM_REGEX = "//(.)\n(.*)";
+    
+    public static int splitAndSum(String text) {
+        return isNullOrEmpty(text) ? 0 : adder(stringsToInts(splitText(text)));
```
삼항 연산자는 다음과 같은 단점들 때문에 사용을 꺼리고 있으니 참고 바람

🛑 삼항 연산자의 주요 단점
1. 🔍 가독성 저하
복잡한 조건식: 조건 자체가 복잡하거나, 참 또는 거짓일 때 실행되는 표현식이 길어지면 한 줄에 많은 내용이 압축되어 코드를 한눈에 이해하기 어려워집니다.

중첩 사용: 삼항 연산자를 중첩해서 사용하면 가독성이 급격히 떨어져 코드를 해독하는 데 시간이 오래 걸립니다. 이는 코드의 의도를 파악하기 어렵게 만들고 버그를 유발할 가능성이 높아집니다.

2. 🐛 디버깅의 어려움
단일 라인 압축: 삼항 연산자는 보통 한 줄로 작성되기 때문에, 조건식이나 결과 표현식에 문제가 발생해도 특정 부분을 분리하여 디버깅하거나 로깅(Logging)하기가 if-else문보다 까다롭습니다.

변수 할당 문제: 조건과 같은 줄에서 변수 할당이 이루어질 때, 예상치 못한 결과가 발생하면 디버깅이 어려워질 수 있습니다.

3. 🧩 제한적인 유연성 (표현식 vs. 문장)
많은 언어에서 삼항 연산자는 **표현식(Expression)**으로 사용되어 값을 반환하는 용도로만 쓰입니다.

따라서 if-else문처럼 **여러 줄의 명령문(Statement)**을 실행하거나, **복잡한 부수 효과(Side Effect)**를 발생시키는 로직을 깔끔하게 처리하기 어렵습니다. 복잡한 로직을 삼항 연산자에 넣으면 코드가 난해해집니다.

4. 📚 학습 곡선 및 언어 의존성
?:와 같은 C 스타일의 문법은 키워드를 사용하는 if-else문보다 추상적이어서 프로그래밍에 익숙하지 않은 사람들에게는 인지 부하를 줄 수 있습니다.

일부 언어(예: Python의 X if Y else Z, Rust의 if가 표현식인 경우)는 삼항 연산자와 유사한 기능을 더 가독성 있게 제공하여, 전통적인 삼항 연산자(?:)의 필요성이 줄어들고 있습니다.

💡 결론 및 사용 권장 사항
삼항 연산자는 매우 간단하고 명확하게 조건에 따라 값을 할당하는 경우(예: int max = (a > b) ? a : b;)에는 코드를 간결하게 만들어 이점이 있습니다.

하지만 위에서 언급된 단점들 때문에, 조금이라도 조건이나 결과 표현식이 복잡해진다고 판단되면 가독성이 좋은 if-else문이나 다른 조건문 구조를 사용하는 것이 유지보수 측면에서 더 권장됩니다.

javajigi commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,48 @@
+    
+    private static String[] splitText(String text) {
+        Matcher m = Pattern.compile(CUSTOM_REGEX).matcher(text);
+        return m.find() ? m.group(2).split(m.group(1)) : text.split(COMMA_AND_COLON_REGEX);
```
삼항 연산자 사용 고려

javajigi commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,48 @@
+        return ints;
+    }
+    
+    private static void negativeNumberCheck(int num) {
+        if(num < 0) {
+            throw new RuntimeException("음수 불가");
+        }
+    }
```
```suggestion
    private static int toInt(String value) {
        int number = Integer.parseInt(value);
        if (number < 0) {
            throw new RuntimeException("음수를 입력할 수 없습니다.");
        }
        return number;
    }
```

위와 같이 구현하는 것은 어떨까?

Mo-bile replied:
@javajigi
이렇게 분리할수도 있군요 제안주셔서 감사합니다. 이 부분 반영하겠습니다!
기계적으로 method 내 indent가 2 이상이면 안될 수 있지만, 이런 경우에는 충분히 가능한 경우로 이해하면 될까요?

또한 제안주신 방식이 '분리한 메소드 안에 또 다른 메서드를 분리한 방식'을 지양할 것을 말씀주셨던 것으로 기억하는데요 이런 경우에 이렇게 해도 괜찮을까요?

물론 저도 정답은 없다고 생각합니다 👍

javajigi reviewed (approved):
피드백에 따라 TDD 사이클을 경험하기 위해 연습한 점이 인상적이네요. 👍 
메서드 또한 한 가지 책임만 가지도록 잘 분리했네요.
삼항 연산자 관련해 피드백 남겼으니 확인해 보시고요.
커밋 관련된 답변엔 채널에 남길께요.

자동차 경주 게임 구현 시작할 때 TDD로 한번 도전해 보는데요.
생각보다 쉽지 않을 수 있습니다.
막막하면 자동차 경주 게임에 대한 기능 목록을 정리한 후에 TDD로 구현 가능한(유틸성 기능)에 대해서만이라도 TDD로 도전해 보세요.
그럼에도 불구하고 막막하다면 포기하고 프로덕션 코드 구현 후 -> 단위 테스트 추가 방식으로 진행해도 좋습니다.

javajigi commented:
Q: 리팩터링 한 기능에 새로운 기능을 구현하면서 깨지는것이 당연한 것 맞을까요?
우선 TDD 싸이클을 겪으면서 느낀 소회를 가볍게 말씀드리면
기능을 만들고 리팩터링하고, 그 다음 순서의 기능을 구현 만들 때 깔끔해진 리팩터링 된 코드가 깨질때 내심 마음이 아팠습니다. 그러면서도 그 구현한 기능을 다시 만들면서 다행이라고 생각들었습니다. 이런 경험이 TDD 싸이클에서 당연한 것일까요?

A: TDD 사이클을 제대로 경험하려면 기능 추가가 먼저가 아니라 실패하는 단위 테스트 코드를 추가한 후에 기능을 추가하는 것이 원칙입니다.
새로운 기능이 추가되면서 코드가 지저분해지는 것은 당연한 결과이니 받아들여야 합니다.
하지만 우리에게는 다음 단계로 리팩터링이 있으니 빠르게 클린 코드로 리팩터링하면 됩니다.
